src/java.base/share/classes/jdk/internal/misc/Unsafe.java

```diff
@@ -2836,6 +2844,50 @@ public final void putDoubleOpaque(Object o, long offset, double x) {
         putDoubleVolatile(o, offset, x);
     }
 
+    private boolean compareAndSetFlatValueAsBytes(Object o, long offset, int layout, Class<?> valueType, Object expected, Object x) {
+        // try nullable atomic array first
+        Object expectedArray = ValueClass.newNullableAtomicArray(valueType, 1);
+        Object xArray = ValueClass.newNullableAtomicArray(valueType, 1);
+        if (arrayLayout(expectedArray.getClass()) != layout) {
```

Comment:
This part feels a bit dirty: we are receiving a request for layout X - we need to be able to create an array with _that same layout_ -- but there's no primitive for doing that. Which means I have to try using different factories and make sure that I get something compatible (which isn't always guaranteed because of VM flags). All this guesswork should not be necessary -- that is, there should be an unsafe API to create an array with the layout I want (assuming that layout is supported of course)

Reply:
I have considered avoiding the array creation -- but that is messy because it means you have to deal with nulls directly. E.g. if either `expected` or `x` are `null`, what is their corresponding numeric configuration? Going through an array element allows the code to remain agnostic as to how nulls are represented.

Reply:
I think runtime support is better for down the road for this part.

Reply:
We have decided to address this by adding a primitive to create a flat array with given known layout:
https://github.com/openjdk/valhalla/pull/1404

Once that API is integrated, I will rebase this PR on top of that.

Reply:
That API is now in. This patch looks fine to me otherwise.

Reply:
I've now simplified the code to use the new primitive.